metronome/checkpointing/service/test/src/io/iohk/metronome/checkpointing/service/CheckpointingServiceProps.scala

```diff
@@ -0,0 +1,407 @@
+      }
+  }
+
+  def run(fixture: TestFixture)(test: TestResources => Task[Prop]): Prop = {
+    import Scheduler.Implicits.global
+
+    fixture.resources.use(test).runSyncUnsafe(timeout = 5.seconds)
+  }
+
+  property("normal execution") = forAll(TestFixture.gen()) { fixture =>
+    run(fixture) { res =>
+      import fixture._
+      import res._
+
+      val execution = batch
+        .map(checkpointingService.executeBlock(_, commitQC, commitPath))
+        .sequence
+
+      val ledgerStorageCheck = store.runReadOnly {
+        ledgerStorage.get(finalLedger.hash)
+      }
+
+      for {
+        results         <- execution
+        persistedLedger <- ledgerStorageCheck
+        ledgerTree      <- ledgerTreeRef.get
+      } yield {
+        val ledgerTreeUpdated = ledgerTree == Map(
+          batch.last.hash -> (finalLedger, batch.last.header)
+        )
+
+        all(
+          "executed correctly" |: results.reduce(_ && _),
+          "ledger persisted" |: persistedLedger.contains(finalLedger),
+          "ledgerTree updated" |: ledgerTreeUpdated
+        )
+      }
+    }
+  }
+
+  property("failed execution - no parent") =
+    forAll(TestFixture.gen(minChain = 2)) { fixture =>
+      run(fixture) { res =>
+        import fixture._
+        import res._
+
+        // parent block or its state is not saved so this must fail
+        val execution = batch.tail
+          .map(checkpointingService.executeBlock(_, commitQC, commitPath))
+          .sequence
+
+        execution
+          .onErrorHandle(_.isInstanceOf[IllegalStateException])
+          .map(_ == true)
+      }
+    }
+
+  property("failed execution - height below last executed") =
+    forAll(TestFixture.gen(minChain = 2)) { fixture =>
+      run(fixture) { res =>
+        import fixture._
+        import res._
+
+        val execution = batch
+          .map(checkpointingService.executeBlock(_, commitQC, commitPath))
+          .sequence
+
+        // repeated execution must fail because we're trying to execute a block of lower height
+        // than the last executed block
+        execution >>
+          execution
+            .onErrorHandle(_.isInstanceOf[IllegalStateException])
+            .map(_ == true)
+      }
+    }
+
+  //TODO: Validate transactions PM-3131/3132
+  //      use a mocked interpreter client that always evaluates blocks as valid
+  property("parallel validation") = forAll(TestFixture.gen(minChain = 4)) {
+    fixture =>
+      run(fixture) { res =>
+        import fixture._
+        import res._
+
+        // validation in random order so blocks need to be persisted first
+        val persistBlocks = store.runReadWrite {
+          batch.map(blockStorage.put).sequence
+        }
+
+        def validation(
+            validating: Ref[Task, Boolean],
+            achievedPar: Ref[Task, Boolean]
+        ) =
+          Task.parSequence {
+            Random
+              .shuffle(batch)
```

Comment:
You can make a seed in `forAll` and use it to create a `Random` instance, that way even the shuffle would be repeatable, in case something fails.

Reply:
Did that in #54